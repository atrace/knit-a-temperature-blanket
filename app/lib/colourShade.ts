export const colourShade = (colour: string, amount: number) => {
  return (
    "#" +
    colour
      .replace(/^#/, "")
      .replace(/../g, (color) =>
        (
          "0" +
          Math.min(255, Math.max(0, parseInt(color, 16) + amount)).toString(16)
        ).slice(-2),
      )
  );
};
