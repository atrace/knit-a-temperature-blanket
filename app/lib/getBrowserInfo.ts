const getBrowserVersion = (
  userAgentString: string,
  browserName: string,
): number | undefined => {
  // Only gets major version
  const versionString = userAgentString.match(
    new RegExp(browserName + "/(\\d+)"),
  );

  if (!versionString || versionString.length > 2) return;

  // Second element is capture group (first element is entire match)
  return parseInt(versionString[1]);
};

export type BrowserName =
  | "firefox"
  | "chrome"
  | "safari"
  | "internet-explorer"
  | "opera"
  | "samsung-internet"
  | "microsoft-edge-legacy"
  | "microsoft-edge-chromium"
  | "unknown";

export const getBrowserInfo = (
  userAgent: string,
): { name: BrowserName; version?: number } => {
  // Based on: https://developer.mozilla.org/en-US/docs/Web/API/Window/navigator#example_1_browser_detect_and_return_a_string
  // and: https://www.geeksforgeeks.org/how-to-detect-the-user-browser-safari-chrome-ie-firefox-and-opera-using-javascript/
  if (userAgent.includes("Firefox")) {
    return {
      name: "firefox",
      version: getBrowserVersion(userAgent, "Firefox"),
    };
  } else if (userAgent.includes("SamsungBrowser")) {
    return {
      name: "samsung-internet",
      version: getBrowserVersion(userAgent, "SamsungBrowser"),
    };
  } else if (userAgent.includes("Opera") || userAgent.includes("OPR")) {
    return {
      name: "opera",
      version: getBrowserVersion(userAgent, "OPR"),
    };
  } else if (userAgent.includes("MSIE") || userAgent.includes("rv:")) {
    return {
      name: "internet-explorer",
      version: getBrowserVersion(userAgent, "MSIE"),
    };
  } else if (userAgent.includes("Edge")) {
    return {
      name: "microsoft-edge-legacy",
      version: getBrowserVersion(userAgent, "Edge"),
    };
  } else if (userAgent.includes("Edg")) {
    return {
      name: "microsoft-edge-chromium",
      version: getBrowserVersion(userAgent, "Edg"),
    };
  } else if (userAgent.includes("Chrome")) {
    return {
      name: "chrome",
      version: getBrowserVersion(userAgent, "Chrome"),
    };
  } else if (userAgent.includes("Safari")) {
    return {
      name: "safari",
      version: getBrowserVersion(userAgent, "Version"),
    };
  } else {
    return { name: "unknown" };
  }
};

type MinimumCompatibleVersion = number;
type BrowserCompatibility =
  | "compatible"
  | "incompatible"
  | MinimumCompatibleVersion;

// Based on: https://caniuse.com/?search=text-wrap on 22/02/24
const textWrapCompatible: { [E in BrowserName]: BrowserCompatibility } = {
  "microsoft-edge-chromium": 114,
  "samsung-internet": 23,
  chrome: 114,
  firefox: 121,
  opera: 100,
  safari: 18,
  "internet-explorer": "incompatible",
  "microsoft-edge-legacy": "incompatible",
  unknown: "incompatible",
};

export const checkBrowserTextWrapCompatible = (userAgent: string): boolean => {
  const browser = getBrowserInfo(userAgent);

  const compatibility = textWrapCompatible[browser.name];

  if (compatibility == "compatible") {
    return true;
  } else if (compatibility == "incompatible") {
    return false;
  } else if (!!browser.version) {
    return browser.version >= compatibility;
  } else {
    return false;
  }
};
