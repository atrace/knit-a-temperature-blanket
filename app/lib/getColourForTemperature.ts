// Ranges are inclusive
export const temperatureRange = [
  { min: -1000, max: -2, colour: "midnightblue" },
  { min: -1, max: 0, colour: "blue" },
  { min: 1, max: 5, colour: "green" },
  { min: 6, max: 10, colour: "greenyellow" },
  { min: 11, max: 15, colour: "lemonchiffon" },
  { min: 16, max: 20, colour: "gold" },
  { min: 21, max: 25, colour: "orange" },
  { min: 26, max: 30, colour: "firebrick" },
  { min: 31, max: 1000, colour: "maroon" }
];

export const getColourForTemperature = (temperature: number) => {
  const integerTemperature = Math.round(temperature);
  let counter = 0;
  while (counter < temperatureRange.length) {
    const range = temperatureRange[counter];
    if (integerTemperature >= range.min && integerTemperature <= range.max) {
      return range.colour;
    }
    counter++;
  }
};
