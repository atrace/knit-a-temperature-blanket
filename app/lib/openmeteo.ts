import { fetchWeatherApi } from "openmeteo";

export interface DailyWeather {
  datetime: Date;
  weatherCode: number; // WMO code
  temperature2mMean: number; // degrees celcius
  daylightDuration: number; // seconds
  precipitationSum: number; // millimetres
}

export const getWeather = async (): Promise<DailyWeather[]> => {
  const startDate = `${new Date().getFullYear()}-01-01`;
  const endDate = new Date().toISOString().slice(0, 10);

  const params = {
    latitude: 53.9566,
    longitude: -1.0774,
    start_date: startDate,
    end_date: endDate,
    daily: [
      "weather_code",
      "temperature_2m_mean",
      "daylight_duration",
      "precipitation_sum"
    ],
    timezone: "Europe/London"
  };
  const url = "https://archive-api.open-meteo.com/v1/archive";
  const responses = await fetchWeatherApi(url, params);

  // Helper function to form time ranges
  const range = (start: number, stop: number, step: number) =>
    Array.from({ length: (stop - start) / step }, (_, i) => start + i * step);

  // Process first location. Add a for-loop for multiple locations or weather models
  const response = responses[0];
  if (!response) return [];

  // Attributes for timezone and location
  const utcOffsetSeconds = response.utcOffsetSeconds();
  const timezone = response.timezone();
  const timezoneAbbreviation = response.timezoneAbbreviation();
  const latitude = response.latitude();
  const longitude = response.longitude();

  const daily = response.daily();
  if (!daily) return [];
 

  // Note: The order of weather variables in the URL query and the indices below need to match!
  const weatherData = {
    daily: {
      time: range(
        Number(daily.time()),
        Number(daily.timeEnd()),
        daily.interval()
      ).map((t) => new Date((t + utcOffsetSeconds) * 1000)),
      weatherCode: daily.variables(0)!.valuesArray()!,
      temperature2mMean: daily.variables(1)!.valuesArray()!,
      daylightDuration: daily.variables(2)!.valuesArray()!,
      precipitationSum: daily.variables(3)!.valuesArray()!
    }
  };

  const dailyWeather: DailyWeather[] = [];

  for (let i = 0; i < weatherData.daily.time.length; i++) {
    dailyWeather.push({
      datetime: weatherData.daily.time[i],
      weatherCode: weatherData.daily.weatherCode[i], // WMO code
      temperature2mMean: weatherData.daily.temperature2mMean[i], // degrees celcius
      daylightDuration: weatherData.daily.daylightDuration[i], // seconds
      precipitationSum: weatherData.daily.precipitationSum[i] // millimetres
    });
  }

  return dailyWeather;
};
