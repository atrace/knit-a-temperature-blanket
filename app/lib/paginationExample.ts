import type { AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";
import axios from "axios";

type PageableResponse = {
  total: number;
  offset: number;
  limit: number;
  results: unknown[];
};

interface ISomeApiClient {
  get<T>(url: string): Promise<AxiosResponse<T>>;
  getWithPagination<T extends PageableResponse>(
    url: string,
    pageSize: number,
    config?: AxiosRequestConfig,
  ): Promise<AxiosResponse<Omit<T, "offset">>>;
}

export class SomeApiClient implements ISomeApiClient {
  baseUrl: string;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  instance: AxiosInstance | null = null;

  get http(): AxiosInstance {
    return this.instance != null ? this.instance : this.initHttp(this.baseUrl);
  }

  initHttp(baseUrl: string) {
    const http = axios.create({
      baseURL: baseUrl,
      withCredentials: true,
      timeout: "20000",
    });

    http.interceptors.request.use((request) => {
      return request;
    });

    http.interceptors.response.use(
      (response) => {
        return response;
      },
      (error) => {
        return Promise.reject(error);
      },
    );

    this.instance = http;
    return http;
  }

  async get<T>(url: string): Promise<AxiosResponse<T>> {
    return await this.http.get<T>(url);
  }

  async getWithPagination<T extends PageableResponse>(
    url: string,
    pageSize = 100,
  ): Promise<AxiosResponse<Omit<T, "offset">>> {
    let response;
    let receivedItems: T["results"] = [];
    const defaultMaxPages = 100;
    let maxPages: number;
    if (process.env.MAX_PAGES_FROM_THIS_API) {
      maxPages =
        parseInt(process.env.MAX_PAGES_FROM_THIS_API) || defaultMaxPages;
    } else {
      maxPages = defaultMaxPages;
    }

    do {
      const newUrl = url.match(/\?/)
        ? `${url}&limit=${pageSize}&offset=${receivedItems.length}`
        : `${url}?limit=${pageSize}&offset=${receivedItems.length}`;

      response = await this.http.get<T>(newUrl);

      if (response.status !== 200) {
        return response;
      }

      if (response.data.total / pageSize > maxPages) {
        throw Error(`Too many pages to paginate: ${response.data.total}`);
      }

      if (response.data.results) {
        receivedItems = receivedItems.concat(response.data.results);
      }
    } while (response.data.total > receivedItems.length);

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { offset, ...responseDataWithoutOffset } = response.data;

    return {
      ...response,
      data: { ...responseDataWithoutOffset, results: receivedItems },
    };
  }
}
