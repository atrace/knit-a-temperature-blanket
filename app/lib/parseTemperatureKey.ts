import {
  TemperatureKey,
  TemperatureRange,
  defaultTemperatureKey,
} from "./getColourForTemperature";

export const parseTemperatureKey = (unparsed: unknown): TemperatureKey => {
  try {
    if (!unparsed || typeof unparsed !== "string" || unparsed.length == 0) {
      throw new Error("no temperature key provided");
    }
    const jsonified = JSON.parse(unparsed);

    if (!Array.isArray(jsonified)) {
      throw new Error("invalid temperature key provided");
    }

    return jsonified.map((element): TemperatureRange => {
      if (
        typeof element["min"] !== "number" ||
        typeof element["max"] !== "number" ||
        element["colour"].length == 0 ||
        element["colourName"].length == 0
      ) {
        throw new Error(`invalid temperature range ${JSON.stringify(element)}`);
      }

      return {
        min: element.min,
        max: element.max,
        colour: element.colour,
        colourName: element.colourName,
      };
    });
  } catch (error) {
    console.error("reverting to default colour key", error);
    return defaultTemperatureKey;
  }
};
