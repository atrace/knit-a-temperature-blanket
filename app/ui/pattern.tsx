import React from "react";
import { getColourForTemperature } from "../lib/getColourForTemperature";
import { DailyWeather } from "../lib/openmeteo";

export default function Pattern({
  dailyWeather
}: {
  dailyWeather: DailyWeather[];
}) {
  //TODO:  Make text item generator generic to allow extra fields to be added easily
  const rows = dailyWeather.map((todaysWeather) => {
    const date = todaysWeather.datetime.toDateString().slice(4, 10);

    const colour = getColourForTemperature(todaysWeather.temperature2mMean);

    // Most recent 2 days come back as NaN :thonk:
    if (!todaysWeather.temperature2mMean) return;

    return (
      <tr key={date}>
        <td>{date}</td>
        <td>({todaysWeather.temperature2mMean.toFixed()}Â°C)</td>

        <td style={{ backgroundColor: colour, width: 200 }}></td>

        <td style={{ marginLeft: 4, display: "flex" }}>
          <p>knit 1 row of</p>
          <p style={{ marginLeft: 4, color: colour }}>{colour}</p>
        </td>
      </tr>
    );
  });

  return (
    <>
      <h2>Date Temperature</h2>
      <table>
        <tbody>{rows}</tbody>
      </table>
    </>
  );
}
