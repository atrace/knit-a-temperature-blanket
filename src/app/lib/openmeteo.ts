import { fetchWeatherApi } from "openmeteo";

export interface DailyWeather {
  datetime: Date;
  weatherCode: number;
  temperature2mMean: number;
  daylightDuration: number;
  precipitationSum: number;
}

export const getWeather = async (): Promise<DailyWeather[]> => {
  const params = {
    latitude: 53.9566,
    longitude: -1.0774,
    start_date: "2024-01-01",
    end_date: "2024-01-18",
    daily: [
      "weather_code",
      "temperature_2m_mean",
      "daylight_duration",
      "precipitation_sum"
    ],
    timezone: "Europe/London"
  };
  const url = "https://archive-api.open-meteo.com/v1/archive";
  const responses = await fetchWeatherApi(url, params);

  // Helper function to form time ranges
  const range = (start: number, stop: number, step: number) =>
    Array.from({ length: (stop - start) / step }, (_, i) => start + i * step);

  // Process first location. Add a for-loop for multiple locations or weather models
  const response = responses[0];

  // Attributes for timezone and location
  const utcOffsetSeconds = response.utcOffsetSeconds();
  const timezone = response.timezone();
  const timezoneAbbreviation = response.timezoneAbbreviation();
  const latitude = response.latitude();
  const longitude = response.longitude();

  const daily = response.daily()!;

  // Note: The order of weather variables in the URL query and the indices below need to match!
  const weatherData = {
    daily: {
      time: range(
        Number(daily.time()),
        Number(daily.timeEnd()),
        daily.interval()
      ).map((t) => new Date((t + utcOffsetSeconds) * 1000)),
      weatherCode: daily.variables(0)!.valuesArray()!,
      temperature2mMean: daily.variables(1)!.valuesArray()!,
      daylightDuration: daily.variables(2)!.valuesArray()!,
      precipitationSum: daily.variables(3)!.valuesArray()!
    }
  };

  const dailyWeather: DailyWeather[] = [];

  for (let i = 0; i < weatherData.daily.time.length; i++) {
    dailyWeather.push({
      datetime: weatherData.daily.time[i],
      weatherCode: weatherData.daily.weatherCode[i], // WMO code
      temperature2mMean: weatherData.daily.temperature2mMean[i], // degrees celcius
      daylightDuration: weatherData.daily.daylightDuration[i], // seconds
      precipitationSum: weatherData.daily.precipitationSum[i] // millimetres
    });
  }

  return dailyWeather;
};
