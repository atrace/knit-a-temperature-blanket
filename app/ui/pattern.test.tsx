import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import Pattern from "./pattern";
import { DailyWeather } from "@/lib/openmeteo";

describe("Pattern", () => {
  it("renders a knitting pattern", () => {
    const dailyWeather: DailyWeather[] = [
      {
        datetime: new Date("04/06/2024"),
        daylightDuration: 3600 * 12,
        weatherCode: 13,
        temperature2mMean: 20,
        precipitationSum: 0
      }
    ];

    const { container } = render(<Pattern dailyWeather={dailyWeather} />);

    expect(container).toMatchSnapshot();
  });
});
