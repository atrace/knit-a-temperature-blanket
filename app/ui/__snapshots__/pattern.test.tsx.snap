// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pattern renders a knitting pattern 1`] = `
<div>
  <h2>
    Date Temperature
  </h2>
  <table>
    <tbody>
      <tr>
        <td>
          Apr 06
        </td>
        <td>
          (
          20
          Â°C)
        </td>
        <td
          style="background-color: gold; width: 200px;"
        />
        <td
          style="margin-left: 4px; display: flex;"
        >
          <p>
            knit 1 row of
          </p>
          <p
            style="margin-left: 4px; color: gold;"
          >
            gold
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
